File: internal/protocol/v1/compatibility.go
package v1

import "strconv"

// MessageKind tells the classifier whether the sender is presenting a control
// or event frame. An unknown wire type cannot safely determine that intent by
// name alone.
type MessageKind string

const (
	MessageKindControl MessageKind = "control"
	MessageKindEvent   MessageKind = "event"
)

type Classification string

const (
	ClassificationKnownControl       Classification = "known_control"
	ClassificationKnownEvent         Classification = "known_event"
	ClassificationUnknownEvent       Classification = "unknown_event"
	ClassificationIncompatible       Classification = "incompatible_version"
	ClassificationUnsupportedControl Classification = "unsupported_control"
)

// Classify applies the v1 compatibility rules without accepting or verifying
// a message. Validation and signature verification remain separate steps.
func Classify(version string, kind MessageKind, messageType string) Classification {
	major, minor, ok := parseVersion(version)
	if !ok || major != 1 {
		return ClassificationIncompatible
	}

	if kind == MessageKindControl {
		if minor != 0 || !IsKnownControl(messageType) {
			return ClassificationUnsupportedControl
		}
		return ClassificationKnownControl
	}
	if kind == MessageKindEvent {
		if IsKnownEvent(messageType) {
			return ClassificationKnownEvent
		}
		return ClassificationUnknownEvent
	}
	return ClassificationIncompatible
}

func IsKnownControl(messageType string) bool {
	for _, known := range KnownControlTypes {
		if string(known) == messageType {
			return true
		}
	}
	return false
}

func IsKnownEvent(messageType string) bool {
	for _, known := range KnownEventTypes {
		if string(known) == messageType {
			return true
		}
	}
	return false
}

func IsTerminalControlResult(status ControlResultStatus) bool {
	return status == ControlResultConfirmed || status == ControlResultRejected || status == ControlResultAmbiguous
}

func FrameWithinLimit(kind MessageKind, size int) bool {
	if size < 0 {
		return false
	}
	switch kind {
	case MessageKindControl:
		return size <= ControlFrameMaxBytes
	case MessageKindEvent:
		return size <= EventFrameMaxBytes
	default:
		return false
	}
}

func parseVersion(version string) (int, int, bool) {
	for i := 1; i < len(version)-1; i++ {
		if version[i] != '.' {
			continue
		}
		major, majorErr := strconv.Atoi(version[:i])
		minor, minorErr := strconv.Atoi(version[i+1:])
		return major, minor, majorErr == nil && minorErr == nil && major >= 0 && minor >= 0
	}
	return 0, 0, false
}

File: web/src/protocol/v1/compatibility.ts
import {
  CONTROL_FRAME_MAX_BYTES,
  CURRENT_VERSION,
  EVENT_FRAME_MAX_BYTES,
  KNOWN_CONTROL_TYPES,
  KNOWN_EVENT_TYPES,
  type ControlResultStatus,
} from "./catalog.generated";

export type MessageKind = "control" | "event";
export type Classification =
  | "known_control"
  | "known_event"
  | "unknown_event"
  | "incompatible_version"
  | "unsupported_control";

const controls = new Set<string>(KNOWN_CONTROL_TYPES);
const events = new Set<string>(KNOWN_EVENT_TYPES);

export function classify(version: string, kind: MessageKind, messageType: string): Classification {
  const parsed = /^(\d+)\.(\d+)$/.exec(version);
  if (!parsed || Number(parsed[1]) !== 1) return "incompatible_version";

  if (kind === "control") {
    if (version !== CURRENT_VERSION || !controls.has(messageType)) return "unsupported_control";
    return "known_control";
  }
  return events.has(messageType) ? "known_event" : "unknown_event";
}

export function isKnownControl(messageType: string): boolean {
  return controls.has(messageType);
}

export function isKnownEvent(messageType: string): boolean {
  return events.has(messageType);
}

export function isTerminalControlResult(status: ControlResultStatus): boolean {
  return status === "confirmed" || status === "rejected" || status === "ambiguous";
}

export function frameWithinLimit(kind: MessageKind, size: number): boolean {
  if (!Number.isSafeInteger(size) || size < 0) return false;
  return size <= (kind === "control" ? CONTROL_FRAME_MAX_BYTES : EVENT_FRAME_MAX_BYTES);
}
